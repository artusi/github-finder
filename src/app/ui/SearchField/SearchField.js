import React, { Component } from "react";
import PropTypes from "prop-types";
import { isNotEmpty } from "ui/helpers/validations";
import SubmitButton from "ui/SubmitButton";
import styles from "./SearchField.module.css";

export class SearchField extends Component {
  constructor(props) {
    super();
    this.state = {
      field: {
        value: props.defaultValue || "",
        valid: false,
        rules: [isNotEmpty]
      }
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    const { value } = event.target;
    const { field } = this.state;
    const valid = field.rules.every(rule => rule(value));

    this.setState({ field: { ...field, value, valid } });
  }

  handleSubmit() {
    if (!this.state.field.value) return;

    this.props.onClick(this.state.field.value);
  }

  render() {
    const { id, label, prefix } = this.props;
    const inputIsEmpty = !this.state.field.value;

    return (
      <div className={styles.searchField}>
        <label className={styles.label} htmlFor={id}>
          {label}
        </label>
        <div className={styles.inputContainer}>
          <span className={styles.prefix}>{prefix}</span>
          <input
            type="text"
            id={`${id}-input`}
            name={id}
            value={this.state.field.value}
            className={styles.fieldInput}
            onChange={this.handleChange}
          />
        </div>
        <SubmitButton
          onClick={this.handleSubmit}
          disabled={inputIsEmpty}
          id="submitSearchField"
        >
          Find!
        </SubmitButton>
      </div>
    );
  }
}

SearchField.propTypes = {
  id: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
  prefix: PropTypes.string,
  defaultValue: PropTypes.string
};

SearchField.defaultProps = {
  defaultValue: "",
  prefix: ""
};

export default SearchField;
