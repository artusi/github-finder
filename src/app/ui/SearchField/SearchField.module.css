.searchField {
  display: flex;
  flex-wrap: wrap;
}
.label {
  margin: 8px 0;
  font-size: 16px;
  font-weight: bold;
  color: #00324d;
}
.inputContainer {
  padding: 6px;
  display: flex;
  flex: 1;
  align-items: center;
}
.prefix {
  font-size: 12px;
  color: #888;
  margin-right: 2px;
}
.fieldInput {
  margin: 0;
  padding: 10px 6px;
  border: 1px solid #d2af48;
  border-radius: 4px;
  flex: 1;
  font-size: 18px;
  width: 140px;
}
