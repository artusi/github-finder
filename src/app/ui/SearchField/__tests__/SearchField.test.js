import React from "react";
import { SearchField } from "../SearchField";

const initialProps = {
  id: "searchTest",
  onClick: jest.fn(),
  label: "Test label for the glory"
};

describe("<SearchField /> - render", () => {
  it("renders", () => {
    const wrapper = shallow(<SearchField {...initialProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("renders with all props > prefix | defaultValue", () => {
    const allProps = {
      ...initialProps,
      prefix: "http://youtube.com/",
      defaultValue: "mkbhd"
    };
    const wrapper = shallow(<SearchField {...allProps} />);

    expect(wrapper).toMatchSnapshot();
  });
});

describe("<SearchField /> - interactions", () => {
  it("Searching for something", () => {
    const simulateClick = jest.fn();

    const wrapper = shallow(
      <SearchField {...initialProps} onClick={simulateClick} />
    );

    wrapper
      .find("#searchTest-input")
      .simulate("change", { target: { name: "searchTest", value: "artusi" } });

    expect(wrapper.state("field").value).toEqual("artusi");

    wrapper.find("#submitSearchField").simulate("click");

    expect(simulateClick).toBeCalled();
  });

  it("Try to search without adding input", () => {
    const simulateClickAgain = jest.fn();

    const wrapper = shallow(
      <SearchField {...initialProps} onClick={simulateClickAgain} />
    );

    wrapper.find("#submitSearchField").simulate("click");

    expect(simulateClickAgain).not.toBeCalled();
  });
});
