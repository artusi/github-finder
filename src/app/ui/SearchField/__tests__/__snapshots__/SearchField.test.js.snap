// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchField /> - render renders 1`] = `
<div
  className="searchField"
>
  <label
    className="label"
    htmlFor="searchTest"
  >
    Test label for the glory
  </label>
  <div
    className="inputContainer"
  >
    <span
      className="prefix"
    />
    <input
      className="fieldInput"
      id="searchTest-input"
      name="searchTest"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <SubmitButton
    customClassName=""
    disabled={true}
    id="submitSearchField"
    onClick={[Function]}
  >
    Find!
  </SubmitButton>
</div>
`;

exports[`<SearchField /> - render renders with all props > prefix | defaultValue 1`] = `
<div
  className="searchField"
>
  <label
    className="label"
    htmlFor="searchTest"
  >
    Test label for the glory
  </label>
  <div
    className="inputContainer"
  >
    <span
      className="prefix"
    >
      http://youtube.com/
    </span>
    <input
      className="fieldInput"
      id="searchTest-input"
      name="searchTest"
      onChange={[Function]}
      type="text"
      value="mkbhd"
    />
  </div>
  <SubmitButton
    customClassName=""
    disabled={false}
    id="submitSearchField"
    onClick={[Function]}
  >
    Find!
  </SubmitButton>
</div>
`;
